File: ios/Sources/DKMadsSSPSDK/Core/CmpConsent.swift
import Foundation

/// IAB CMP storage (UMP / TCF / USP / GPP) — same keys as Android `CmpConsent.kt`.
struct CmpSnapshot {
    var tcfString: String?
    var gdprApplies: Bool?
    var uspString: String?
    var gppString: String?
    var gppSid: String?
}

enum CmpConsent {
    private static let tcfKeys = ["IABTCF_TCString", "IABTCF_ConsentString"]
    private static let gdprKeys = ["IABTCF_gdprApplies"]
    private static let uspKeys = ["IABUSPrivacy_String"]
    private static let gppStringKeys = ["IABGPP_GppString", "IABGPP_HDR_GppString"]
    private static let gppSidKeys = ["IABGPP_SID", "IABGPP_SectionId"]

    static func readSnapshot() -> CmpSnapshot {
        let defaults = UserDefaults.standard
        let tcf = firstNonEmpty(defaults, keys: tcfKeys)
        let gdprRaw = firstInt(defaults, keys: gdprKeys)
        let gdpr: Bool? = gdprRaw.map { $0 == 1 }
        return CmpSnapshot(
            tcfString: tcf,
            gdprApplies: gdpr,
            uspString: firstNonEmpty(defaults, keys: uspKeys),
            gppString: firstNonEmpty(defaults, keys: gppStringKeys),
            gppSid: firstNonEmpty(defaults, keys: gppSidKeys)
        )
    }

    static func merge(into existing: ConsentData) -> ConsentData {
        let snap = readSnapshot()
        var out = existing
        if (out.consentString ?? "").trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
           let tcf = snap.tcfString, !tcf.isEmpty {
            out.consentString = tcf
        }
        if let gdpr = snap.gdprApplies { out.gdpr = out.gdpr || gdpr }
        if (out.usPrivacyString ?? "").trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
           let usp = snap.uspString, !usp.isEmpty {
            out.usPrivacyString = usp
            out.ccpa = out.ccpa || true
        }
        if (out.gppString ?? "").trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
           let gpp = snap.gppString, !gpp.isEmpty {
            out.gppString = gpp
        }
        if (out.gppSid ?? "").trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
           let sid = snap.gppSid, !sid.isEmpty {
            out.gppSid = sid
        }
        if out.attStatus == nil {
            out.attStatus = AdvertisingIdentifiers.attStatus()
        }
        return out
    }

    static func hasMinimalConsent(_ consent: ConsentData) -> Bool {
        if !(consent.consentString ?? "").trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { return true }
        if consent.resolvedUsPrivacyString() != nil { return true }
        if !(consent.gppString ?? "").trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { return true }
        return false
    }

    private static func firstNonEmpty(_ defaults: UserDefaults, keys: [String]) -> String? {
        for key in keys {
            let v = defaults.string(forKey: key)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            if !v.isEmpty { return v }
        }
        return nil
    }

    private static func firstInt(_ defaults: UserDefaults, keys: [String]) -> Int? {
        for key in keys {
            if defaults.object(forKey: key) != nil {
                return defaults.integer(forKey: key)
            }
        }
        return nil
    }
}

File: android/CmpConsent.kt
package com.dkmads.ssp

import android.content.Context

/**
 * Read IAB TCF / USP / GPP strings from CMP SDK storage (Sourcepoint, OneTrust, in-app CMP, etc.).
 * Explicit [SSPSDK.setConsent] values take precedence over CMP when non-blank.
 */
data class CmpSnapshot(
    val tcfString: String? = null,
    val gdprApplies: Boolean? = null,
    val uspString: String? = null,
    val gppString: String? = null,
    val gppSid: String? = null,
)

internal object CmpConsent {
    private val tcfKeys = listOf("IABTCF_TCString", "IABTCF_ConsentString")
    private val gdprKeys = listOf("IABTCF_gdprApplies")
    private val uspKeys = listOf("IABUSPrivacy_String")
    private val gppStringKeys = listOf("IABGPP_GppString", "IABGPP_HDR_GppString")
    private val gppSidKeys = listOf("IABGPP_SID", "IABGPP_SectionId")

    fun readSnapshot(context: Context): CmpSnapshot {
        val app = context.applicationContext
        val prefNames = listOf(
            "IABTCF",
            "${app.packageName}_preferences",
            "${app.packageName}.preferences",
            "com.google.android.gms.measurement.prefs",
        )
        var tcf: String? = null
        var gdpr: Boolean? = null
        var usp: String? = null
        var gpp: String? = null
        var sid: String? = null
        for (name in prefNames) {
            try {
                val prefs = app.getSharedPreferences(name, Context.MODE_PRIVATE)
                if (tcf.isNullOrBlank()) tcf = firstString(prefs, tcfKeys)
                if (gdpr == null) gdpr = firstGdpr(prefs)
                if (usp.isNullOrBlank()) usp = firstString(prefs, uspKeys)
                if (gpp.isNullOrBlank()) gpp = firstString(prefs, gppStringKeys)
                if (sid.isNullOrBlank()) sid = firstString(prefs, gppSidKeys)
            } catch (_: Throwable) {
                /* optional CMP */
            }
        }
        return CmpSnapshot(
            tcfString = tcf?.trim()?.takeIf { it.isNotEmpty() },
            gdprApplies = gdpr,
            uspString = usp?.trim()?.takeIf { it.isNotEmpty() },
            gppString = gpp?.trim()?.takeIf { it.isNotEmpty() },
            gppSid = sid?.trim()?.takeIf { it.isNotEmpty() },
        )
    }

    fun mergeInto(existing: ConsentState, snap: CmpSnapshot): ConsentState {
        return ConsentState(
            gdpr = existing.gdpr || (snap.gdprApplies == true),
            ccpa = existing.ccpa || !snap.uspString.isNullOrBlank(),
            consentString = existing.consentString?.trim()?.takeIf { it.isNotEmpty() } ?: snap.tcfString,
            gppString = existing.gppString?.trim()?.takeIf { it.isNotEmpty() } ?: snap.gppString,
            gppSid = existing.gppSid?.trim()?.takeIf { it.isNotEmpty() } ?: snap.gppSid,
            usPrivacyString = existing.usPrivacyString?.trim()?.takeIf { it.isNotEmpty() } ?: snap.uspString,
        )
    }

    fun hasMinimalConsent(state: ConsentState): Boolean {
        if (!state.consentString.isNullOrBlank()) return true
        if (state.resolvedUsPrivacyString() != null) return true
        if (!state.gppString.isNullOrBlank()) return true
        return false
    }

    /** @deprecated use [readSnapshot] */
    fun readUspString(context: Context): String? = readSnapshot(context).uspString

    private fun firstString(prefs: android.content.SharedPreferences, keys: List<String>): String? {
        for (key in keys) {
            val v = prefs.getString(key, null)?.trim()
            if (!v.isNullOrEmpty()) return v
        }
        return null
    }

    private fun firstGdpr(prefs: android.content.SharedPreferences): Boolean? {
        for (key in gdprKeys) {
            if (!prefs.contains(key)) continue
            return prefs.getInt(key, 0) == 1
        }
        return null
    }
}
